### PR TITLE
[PATCH] add log file rotation

### PR DESCRIPTION
**Pull request by sayanibhattacharjee**
Co-Authored-By: rivencxl <smartcxl@gmail.com>
@rivencxl I have rebased your patch, please take a look.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Add support for log_file rotation.
```

### CODE REVIEW

File: klog.go [hunk 1]
```diff
@@ -424,10 +426,12 @@ func InitFlags(flagset *flag.FlagSet) {
 	}
 
 	flagset.StringVar(&logging.logDir, "log_dir", logging.logDir, "If non-empty, write log files in this directory")
-	flagset.StringVar(&logging.logFile, "log_file", logging.logFile, "If non-empty, use this log file")
+	flagset.StringVar(&logging.logFile, "log_file", logging.logFile, "If non-empty, use this log file (mutually exclusive with the log-dir option)")
 	flagset.Uint64Var(&logging.logFileMaxSizeMB, "log_file_max_size", logging.logFileMaxSizeMB,
 		"Defines the maximum size a log file can grow to. Unit is megabytes. "+
 			"If the value is 0, the maximum file size is unlimited.")
+	flagset.IntVar(&logging.logFileMaxBackups, "log_file_max_backups", logging.logFileMaxBackups,
```
Comment (dims) [inline klog.go:433]: `"log_file_max_backups"` should be `"klog_log_file_max_backups"` here @sayanibhattacharjee

Comment (sayanibhattacharjee) [reply klog.go]: Fixed this.

File: klog_test.go [hunk 2]
```diff
@@ -1398,19 +1398,20 @@ func (l *testLogr) WithValues(...interface{}) logr.Logger {
...
-	"skip_log_headers":  {},
-	"stderrthreshold":   {},
-	"v":                 {},
-	"vmodule":           {},
+	"add_dir_header":            {},
+	"alsologtostderr":           {},
+	"log_backtrace_at":          {},
+	"log_file":                  {},
+	"log_file_max_size":         {},
+	"klog_log_file_max_backups": {},
```
Comment (yuzhiquan) [inline klog_test.go:1406]: New flag start with klog no need to be added in this map.

File: klog_file.go [hunk 3]
```diff
@@ -116,6 +117,18 @@ func logName(tag string, t time.Time) (name, link string) {
 	return name, program + "." + tag
 }
 
+func rotateLogName(logFile string, index uint64, t time.Time) string {
+	return fmt.Sprintf("%s.%04d%02d%02d-%02d%02d%02d.%d",
```
Comment (yuzhiquan) [inline klog_file.go:121]: How about to add new flag set rotate log file time format?  @dims

File: klog_test.go (hunk 2 above)
Comment (dims) [reply klog_test.go:1406]: correct, let's drop this @sayanibhattacharjee

File: klog_file.go (hunk 3 above)
Comment (dims) [reply klog_file.go:121]: @yuzhiquan we can do that in a follow up?

Comment (yuzhiquan) [reply klog_file.go:121]: > @yuzhiquan we can do that in a follow up?

Sure, it's useful for someone,like me,to set specified time format for rotate log file, but i do not known it's helpful or not?

**Comment by xuzhenglun** [comment]
HI @yuzhiquan :

I notice that this pr had been pending 3 months. any updated?

**Comment by yuzhiquan** [comment]
> HI @yuzhiquan :
> 
> I notice that this pr had been pending 3 months. any update?

@sayanibhattacharjee @dims

**Comment by yuzhiquan** [comment]
@sayanibhattacharjee plz squash.
@dims more review comments?

**Pull request closed without merging**